File: backend/app/routes/newsletter.py
from fastapi import APIRouter, HTTPException
from app.supabase_client import supabase
from app.schemas.newsletter import (
    NewsletterCreate,
    NewsletterResponse,
)

router = APIRouter(
    prefix="/newsletter",
    tags=["Newsletter"],
)

@router.post("/", response_model=NewsletterResponse)
def subscribe_newsletter(
    newsletter: NewsletterCreate,
):
    # Query supabase
    response = (
        supabase.table("newsletter_subscribers")
        .select("*")
        .eq("email", newsletter.email)
        .execute()
    )

    if response.data:
        raise HTTPException(
            status_code=409,
            detail="Email already subscribed."
        )

    # Insert subscriber
    insert_response = (
        supabase.table("newsletter_subscribers")
        .insert({"email": newsletter.email})
        .execute()
    )

    if not insert_response.data:
        raise HTTPException(
            status_code=500,
            detail="Failed to subscribe email."
        )

    return insert_response.data[0]

File: backend/src/routes/newsletter.ts
import { Router, Request, Response } from "express";
import { z } from "zod";
import { supabase } from "../supabase.js";

export const newsletterRouter = Router();

const newsletterSchema = z.object({
  email: z.string().email(),
});

// POST /newsletter/
newsletterRouter.post("/", async (req: Request, res: Response): Promise<void> => {
  try {
    const validatedBody = newsletterSchema.parse(req.body);

    // Query if already subscribed
    const { data: existing, error: selectError } = await supabase
      .from("newsletter_subscribers")
      .select("*")
      .eq("email", validatedBody.email);

    if (selectError) {
      res.status(500).json({ detail: selectError.message });
      return;
    }

    if (existing && existing.length > 0) {
      res.status(409).json({ detail: "Email already subscribed." });
      return;
    }

    // Insert subscriber
    const { data, error } = await supabase
      .from("newsletter_subscribers")
      .insert({ email: validatedBody.email })
      .select();

    if (error || !data || data.length === 0) {
      res.status(500).json({ detail: error?.message || "Failed to subscribe email." });
      return;
    }

    res.status(200).json(data[0]);
  } catch (error: any) {
    if (error instanceof z.ZodError) {
      res.status(422).json({ detail: error.errors });
      return;
    }
    res.status(500).json({ detail: error.message || "Internal server error" });
  }
});
